spring-batch-integration/src/main/java/org/springframework/batch/integration/chunk/RemoteChunkingMasterStepBuilder.java

```diff
@@ -0,0 +1,426 @@
+import org.springframework.batch.item.ItemStream;
+import org.springframework.batch.item.ItemWriter;
+import org.springframework.batch.repeat.CompletionPolicy;
+import org.springframework.batch.repeat.RepeatOperations;
+import org.springframework.batch.repeat.exception.ExceptionHandler;
+import org.springframework.beans.factory.config.BeanDefinition;
+import org.springframework.beans.factory.support.BeanDefinitionBuilder;
+import org.springframework.context.support.GenericApplicationContext;
+import org.springframework.integration.channel.DirectChannel;
+import org.springframework.integration.core.MessagingTemplate;
+import org.springframework.messaging.PollableChannel;
+import org.springframework.retry.RetryPolicy;
+import org.springframework.retry.backoff.BackOffPolicy;
+import org.springframework.retry.policy.RetryContextCache;
+import org.springframework.transaction.PlatformTransactionManager;
+import org.springframework.transaction.interceptor.TransactionAttribute;
+import org.springframework.util.Assert;
+
+import java.util.function.Function;
+
+/**
+ * Builder for a master step in a remote chunking setup. This builder:
+ *
+ * <ul>
+ *     <li>creates and sets a {@link ChunkMessageChannelItemWriter} on the master step</li>
+ *     <li>registers a {@link RemoteChunkHandlerFactoryBean} in the given context</li>
+ * </ul>
+ *
+ * @param <I> type of input items
+ * @param <O> type of output items
+ *
+ * @since 4.1
+ * @author Mahmoud Ben Hassine
+ */
+public class RemoteChunkingMasterStepBuilder<I, O> extends FaultTolerantStepBuilder<I, O> {
+
+	private MessagingTemplate messagingTemplate;
+	private GenericApplicationContext applicationContext;
+	private PollableChannel inputChannel;
+	private DirectChannel outputChannel;
+
+	private final int DEFAULT_MAX_WAIT_TIMEOUTS = 40;
+	private static final long DEFAULT_THROTTLE_LIMIT = 6;
+	private int maxWaitTimeouts = DEFAULT_MAX_WAIT_TIMEOUTS;
+	private long throttleLimit = DEFAULT_THROTTLE_LIMIT;
+
+	/**
+	 * Create a new {@link RemoteChunkingMasterStepBuilder}.
+	 *
+	 * @param stepName name of the master step
+	 */
+	public RemoteChunkingMasterStepBuilder(String stepName) {
+		super(new StepBuilder(stepName));
+	}
+
+	/**
+	 * Set the application context in which required beans will be registered.
+	 *
+	 * @param applicationContext the application context to set.
+	 * @return this builder instance for fluent chaining
+	 */
+	public RemoteChunkingMasterStepBuilder<I, O> applicationContext(GenericApplicationContext applicationContext) {
+		Assert.notNull(applicationContext, "applicationContext must not be null");
+		this.applicationContext = applicationContext;
+		return this;
+	}
+
+	/**
+	 * Set the input channel on which replies from workers will be received.
+	 *
+	 * @param inputChannel the input channel
+	 * @return this builder instance for fluent chaining
+	 */
+	public RemoteChunkingMasterStepBuilder<I, O> inputChannel(PollableChannel inputChannel) {
```

Comment:
Any of these setters that map to something else, I try to put a `@see` on as well.  For this example, I'd do `@see ChunkMessageChannelItemWriter#setReplyChannel()`.

Reply:
fixed